OPEN-74 fix method and super-class generics

emanuel-persequor opened the pull request.

emanuel-persequor committed:
OPEN-74 fix method and super-class generics

emanuel-persequor committed:
OPEN-74 fall back to Object as return type when unknown

This may not be the correct solution, but it's unclear how else
to handle this

emanuel-persequor committed:
OPEN-74 Has(Class/Method) level generic methods and safer defaults

emanuel-persequor committed:
OPEN-74 Handle parameterized types too

emanuel-persequor committed:
OPEN-74 Break infinite loop when generics extends themselves

emanuel-persequor commented:
@anton-larsen-persequor I don't understand why the build-pipelines fail. Also, I have two FIXME's which go beyond my understanding of Generics in ran. But this PR fixes a number of challenges with method generics, generics in hierarchies, etc.

anton-larsen-persequor commented on ran-core/src/main/java/io/ran/ClazzMethod.java:
```diff
@@ -83,8 +90,67 @@ public boolean isSynchronized() {
+
 	public Clazz<?> getReturnType() {
-		return Clazz.of(method.getReturnType());
+		Class<?> declaringClass = method.getDeclaringClass();
```
I don't see any tests that show that getReturnType still works for non generic types, which the implementation of this suggest may not work correctly. Considering it is used a lot of other places it may have been tested implicitly, but an explicit test would be better to have as well

emanuel-persequor committed:
OPEN-74 Added missing test for non-generic methods

emanuel-persequor committed:
OPEN-74 Handle parameterized types in parameters too

emanuel-persequor committed:
OPEN-74 Dont set generic clazz if its the same as clazz

emanuel-persequor committed:
OPEN-74 Remove redundant genericClazz field

This information is now captured by the clazz field of the
parameter, holding both simple type/class info and generic
info in one place

Commit:
OPEN-74 added test for bound wildcards in method parameters and return values

emanuel-persequor committed:
OPEN-74 Handle <? super U> case as well

emanuel-persequor committed:
OPEN-74 More testing

emanuel-persequor committed:
OPEN-74 Handle bound generic return types better

emanuel-persequor committed:
OPEN-74 Ensure generics are set on interfaces

emanuel-persequor committed:
OPEN-74 Combine bounds of class with wildcard generic

This is one more case in which a method has wildcard, but the
type has bounds which are more strict.

emanuel-persequor committed:
OPEN-74 Even more generic cases

emanuel-persequor committed:
OPEN-74 Add isArray and Array<=>Component type juggling

emanuel-persequor committed:
OPEN-74 Boolean is not an int

emanuel-persequor committed:
OPEN-74 Support array-types

MikePSQR committed:
SAGA-7131 loopy loop

MikePSQR committed:
SAGA-7131 findGenericSuper returns map of generics from super to this

MikePSQR committed:
SAGA-7131 fancy most specific but gets looped

MikePSQR committed:
SAGA-7131 cleanup and tests

MikePSQR committed:
SAGA-7131 loop headache

MikePSQR committed:
SAGA-7131 cleaned up with no loops in generics

MikePSQR committed:
moved tests around a bit and fixed some overflows

MikePSQR committed:
formatting

emanuel-persequor committed:
OPEN-74 Handle NPE when finding methods of Object

emanuel-persequor committed:
OPEN-74 default for void is null

This may not make a lot of sense, but the alternative was
throwing an exception which has worse effects.

MikePSQR committed:
fixed ignored test, just in case

MikePSQR committed:
finished merge from emanuel

MikePSQR commented:
Reworked in #48

emanuel-persequor committed:
OPEN-74 Added failing regression tests

MikePSQR committed:
fixed emanuels tests, LOOP_STOP needs to be redone

MikePSQR committed:
hack fix

emanuel-persequor committed:
OPEN-74 Add methods of interfaces unless overridden

emanuel-persequor committed:
OPEN-74 Make the test a bit more elaborate (Weird)

emanuel-persequor committed:
OPEN-74 With raw and bounds, bounds should be taken into account

MikePSQR committed:
potential test for future. todo rework LOOP_STOP to take into account defining class, not just name of type variable. add tests for raw classes

emanuel-persequor committed:
OPEN-74 Disable test which is does not reflect the current behaviour

emanuel-persequor committed:
Merge pull request #48 from persequor-com/bugfix/OPEN-74-mike

OPEN-74 Rework

MikePSQR committed:
fix for loop stop, todo more raw class tests

MikePSQR committed:
OPEN-74 added some tests with raw generic classes, re-added coverage field pattern check

MikePSQR committed:
OPEN-74 warning

MikePSQR committed:
Merge branch 'main' into bugfix/OPEN-74-method-generics-improved

niels-psqr commented on ran-core/src/test/java/io/ran/ClazzMethodRegression.java:
Lots of old comments that needs to be removed

MikePSQR replied:
done

niels-psqr commented on ran-core/src/test/java/io/ran/ClazzMethodTest.java:
```diff
@@ -0,0 +1,535 @@
+	public void testHasGenericFromMethod_parameterized() {
+		// <T2> List<T2> param2(List<T2> input)
+		ClazzMethod param2 = Clazz.of(GenericTesterImpl.class).methods().find("param2", List.class, List.class).get();
+		assertMethod(g(GenericTester.class, String.class), param2, g(List.class, Object.class), g(List.class, Object.class));
```
Should these lists not contain String.class? Or is that erased by the compiler?

niels-psqr commented on ran-core/src/test/java/io/ran/ClazzMethodTest.java:
```diff
@@ -0,0 +1,535 @@
+		// <T2 extends T> List<T2> mixed3(List<T2> input)
+		ClazzMethod mixed3 = Clazz.of(GenericTesterImpl.class).methods().find("mixed3", List.class, List.class).get();
+		assertMethod(g(GenericTester.class, String.class), mixed3, g(List.class, String.class), g(List.class, String.class));
+		assertFalse(mixed3.hasGenericFromClass());
```
Should this not be true?

MikePSQR replied:
that's debatable, I could try to make it so

MikePSQR committed:
OPEN-74 formatting and comments

MikePSQR committed:
OPEN-74 formatting and comments

niels-psqr commented on ran-core/src/test/java/io/ran/ClazzMethodTest.java:
```diff
@@ -0,0 +1,535 @@
+	}
+
+	@Test
+	public void testICassandraEventsQuery_ReturnCase_fromWildCard() {
```
This test contains the above test by accident?

MikePSQR replied:
merged them and used assertMethod

MikePSQR committed:
OPEN-74 formatting

niels-psqr commented on ran-core/src/test/java/io/ran/ClazzMethodTest.java:
```diff
@@ -0,0 +1,535 @@
+	public void testMoreSpecificDifferentBounds() {
+		// ListHolder<? extends Set<String>> make(); ListHolder<F extends List<?>>
+		// this is effectively multiple bounds, not supported
+		assertThrows(IllegalArgumentException.class,
+				() -> Clazz.of(StringSetHolderFactory.class).methods().find("make", ListHolder.class));
```
Not sure Iunderstand why this is an illegal operation

MikePSQR replied:
it is effectively multiple bound generics, which we don't support. The method returns ListHolder of something that is both a Set<String> and List<?>

niels-psqr commented on ran-core/src/main/java/io/ran/Clazz.java:
```diff
@@ -35,16 +41,132 @@
+			throw new IllegalArgumentException("multiple bounds are not supported " + typeVariable);
+		}
+
+		String fullName = typeVariable.getGenericDeclaration().toString() + " " + typeVariable.getName();
```
Relying on the toString declarition seems like a bad idea? Honestly not sure if the method is something that can be changed easily by accident.

niels-psqr replied:
Mostly considering if potentially the toString can include memory references, making it a not match when it should

MikePSQR replied:
:man_shrugging:

niels-psqr commented on ran-core/src/main/java/io/ran/ClazzMethod.java:
```diff
@@ -44,7 +53,7 @@ public <T extends Annotation> T getAnnotation(Class<T> annotation) {
 
 	public boolean matches(String token) {
-		return this.methodToken.equals(token);
+		return method.toString().equals(token);
```
Again a toString, not sure this is 100% stable, through it probably is

MikePSQR replied:
it used to be the same thing, just the toString() was called in constructor

niels-psqr commented on ran-core/src/main/java/io/ran/ClazzMethod.java:
```diff
@@ -83,12 +92,61 @@ public boolean isSynchronized() {
 	}
 
+	private List<Type> getAllGenericTypes() {
+		ArrayList<Type> types = new ArrayList<>();
+		types.add(method.getGenericReturnType());
+		types.addAll(Arrays.asList(method.getGenericParameterTypes()));
+		types.addAll(Arrays.asList(method.getGenericExceptionTypes()));
+		return types.stream().flatMap(this::flattenGenericTypes).collect(Collectors.toList());
```
Should this maybe be a set? Since a generic type can appear in multiple places but the caller of the method has no way to differentiate them in the return

MikePSQR replied:
done

MikePSQR committed:
OPEN-74 consistent handling of raw classes / interfaces. now they get default generics

niels-psqr commented on ran-core/src/main/java/io/ran/ClazzMethod.java:
```diff
@@ -83,12 +92,61 @@ public boolean isSynchronized() {
+			if (type instanceof TypeVariable) {
+				if (((TypeVariable<?>) type).getGenericDeclaration() == method)
+					return true;
+			} else if (type instanceof WildcardType) {
+				return true;
+			}
```
Should mention in the JavaDocs that wildcards are always treated as method-level generic

MikePSQR replied:
what javadocs :laughing:

MikePSQR replied:
done

niels-psqr reviewed (commented):
Left some general comments.

As we talked about the code base in general needs better test coverage, that actually tests the code and not just runs it. With that being said, that is out of scope for this PR. I've created #49 that adds PIT on all PRs on the repo, which should help us to get an overview of how test coverage is doing, albeit the reports are not that well presented yet.

Overall this is probably ready to be merged.

MikePSQR committed:
OPEN-74 some test cases rewritten

MikePSQR committed:
OPEN-74 method generics handled as set

MikePSQR committed:
OPEN-74 javadoc

niels-psqr commented on ran-core/src/main/java/io/ran/Clazz.java:
```diff
@@ -413,31 +527,40 @@ public Type getOwnerType() {
+	 * @return Clazz of `ofClazz` with generics set as specified by `this`
+	 */
 	Clazz<?> findGenericSuper(Class<?> ofClazz) {
+		if (ofClazz.equals(Object.class)) { // todo I added this, is it an issue?
```
@todo: This should be safe to add, as far as I can tell

MikePSQR replied:
:+1:

niels-psqr commented on ran-core/src/main/java/io/ran/Clazz.java:
```diff
@@ -413,31 +527,40 @@ public Type getOwnerType() {
 	}
 
-	public boolean equals(Clazz<?> clazz) {
+	public boolean equals(Clazz<?> clazz) { // todo what about other equals
```
@todo: Not sure what this TODO is refering to

MikePSQR committed:
OPEN-74 solved some todos, mainly enforcing correct generics when creating a Clazz

MikePSQR committed:
OPEN-74 deprecate weird .equals()

MikePSQR committed:
OPEN-74 removed todos, added todos

MikePSQR committed:
OPEN-74 little cleanup

MikePSQR committed:
OPEN-74 untodod todo

MikePSQR committed:
OPEN-74 mini cleanup

niels-psqr reviewed (approved):
At this point I think this is ready to be merged.

It should overall be safe & functional. We do need to watch out for any potential performance degradations though, but those will only really be visible in the bigger picture, if they are present at all.
